File: LEDS/ledcontroller.py
import time
import math

from ledstrip import LEDStrip
import os

class LEDController:
    def __init__(self, clk_pin, dat_pin, debug=False):
        self.debug = debug
        self.MODES = {"STATIC_COLOR": 1, "COLOR_CYCLE": 2, "BREATHING": 3}
        self.mode = self.MODES["STATIC_COLOR"]  # Initial mode

        # Color definitions
        self.static_color = [255, 20, 147]  # DeepPink
        self.cycle_colors = [[255, 87, 51], [227, 11, 92], [64, 224, 208]]
        self.color_index = 0
        self.color_change_interval = 5  # Interval in seconds
        self.last_color_change_time = time.time()

        # Breathing mode
        self.breath_cycle_duration = 7  # Duration in seconds

        # Flag to track if static color has been set
        self.color_set = False

        # Only initialize the LEDStrip if not in debug mode
        if not self.debug:
            self.strip = LEDStrip(clk_pin, dat_pin)
        else:
            print("Debug mode active: LEDStrip hardware initialization skipped.")

    def switch_mode(self):
        self.mode = (self.mode % 3) + 1
        print(f"Mode switched to {'STATIC_COLOR' if self.mode == self.MODES['STATIC_COLOR'] else 'COLOR_CYCLE' if self.mode == self.MODES['COLOR_CYCLE'] else 'BREATHING'}.")

    def static_color_mode(self):
        if not self.color_set:
            hex_color = "#{:02x}{:02x}{:02x}".format(*self.static_color)
            print("\033[48;2;{};{};{}m Static color set to: {} \033[0m".format(*self.static_color, hex_color))
            if not self.debug:
                self.strip.setcolourrgb(*self.static_color)
            self.color_set = True

    def color_cycle_mode(self):
        current_time = time.time()
        if current_time - self.last_color_change_time > self.color_change_interval:
            self.last_color_change_time = current_time
            start_color = self.cycle_colors[self.color_index]
            self.color_index = (self.color_index + 1) % len(self.cycle_colors)
            end_color = self.cycle_colors[self.color_index]
            duration = 5  # Duration of color transition in seconds
            transition_interval = 0.05  # Interval for color transition in seconds
            steps = int(duration / transition_interval)
            for step in range(steps + 1):
                ratio = step / steps
                transition_color = [int(start + (end - start) * ratio) for start, end in zip(start_color, end_color)]
                hex_color = "#{:02x}{:02x}{:02x}".format(*transition_color)
                print("\033[48;2;{};{};{}m Color transitioned to: {} \033[0m".format(*transition_color, hex_color))
                if not self.debug:
                    self.strip.setcolourrgb(*transition_color)
                time.sleep(transition_interval)

    def breathing_mode(self):
        current_time = time.time()
        time_into_cycle = (current_time % self.breath_cycle_duration) / self.breath_cycle_duration
        brightness = (math.sin(time_into_cycle * 2 * math.pi) + 1) / 2
        adjusted_color = [int(c * brightness) for c in self.cycle_colors[0]]
        hex_color = "#{:02x}{:02x}{:02x}".format(*adjusted_color)
        print("\033[48;2;{};{};{}m Breathing mode: brightness {:.2f}, color: {} \033[0m".format(*adjusted_color, brightness, hex_color))
        if not self.debug:
            self.strip.setcolourrgb(*adjusted_color)

    def run(self):
        try:
            while True:
                # check if a temp file exists to change mode
                try:
                    with open("temp.txt", "r") as f:
                        self.switch_mode()
                        f.close()
                        # remove the temp file
                        os.remove("temp.txt")
                except FileNotFoundError:
                    pass

                if self.mode == self.MODES["STATIC_COLOR"]:
                    self.static_color_mode()
                    time.sleep(1)
                elif self.mode == self.MODES["COLOR_CYCLE"]:
                    self.color_cycle_mode()
                    time.sleep(1)
                elif self.mode == self.MODES["BREATHING"]:
                    self.breathing_mode()
                    time.sleep(0.1)
        except KeyboardInterrupt:
            print("Program exited gracefully")
        finally:
            if not self.debug:
                self.strip.cleanup()

File: same_node_js/ledcontroller.js
const LEDStrip = require('./ledstrip.js'); // Import the LEDStrip class from your custom module
const chalk = require('chalk');

class LEDController {
    constructor(clockPin, dataPin, debug = false) {
        this.debug = debug;
        this.MODES = { STATIC_COLOR: 1, COLOR_CYCLE: 2, BREATHING: 3 };
        this.mode = this.MODES.STATIC_COLOR; // Initial mode

        // Color definitions
        this.staticColor = [255, 20, 147]; // DeepPink
        this.cycleColors = [[255, 87, 51], [227, 11, 92], [64, 224, 208]];
        this.colorIndex = 0;
        this.colorChangeInterval = 5000; // Interval in milliseconds
        this.lastColorChangeTime = Date.now();

        // Breathing mode
        this.breathCycleDuration = 7000; // Duration in milliseconds

        this.interval = null;

        // Flag to track if static color has been set
        this.colorSet = false;

        // Only initialize the LEDStrip if not in debug mode
        if (!this.debug) {
            this.strip = new LEDStrip(clockPin, dataPin);
        } else {
            console.log("Debug mode active: LEDStrip hardware initialization skipped.");
        }
    }

    switchMode() {
        this.mode = (this.mode % 3) + 1;
        console.log(`Mode switched to ${this.mode === this.MODES.STATIC_COLOR ? 'STATIC_COLOR' : this.mode === this.MODES.COLOR_CYCLE ? 'COLOR_CYCLE' : 'BREATHING'}.`);
    }

    staticColorMode() {
        // console.log("Executing Static Color Mode.");
        // console.log(`Static color set to: ${this.staticColor}.`);
        // if (!this.debug) {
        //     this.strip.setColourRGB(this.staticColor[0], this.staticColor[1], this.staticColor[2]);
        // }

        // console.log("Executing Static Color Mode.");
        if (!this.colorSet) {
            const hexColor = `#${this.staticColor.map(c => c.toString(16).padStart(2, '0')).join('')}`;
            console.log(chalk.bgHex(hexColor).white(` Static color set to: ${hexColor} `));
            if (!this.debug) {
                this.strip.setColourRGB(this.staticColor[0], this.staticColor[1], this.staticColor[2]);
            }
            this.colorSet = true;
        } else {
            // console.log("Static color mode already set. Skipping.");
        }
    }

    // colorCycleMode() {
    //     this.colorSet = false;
    //     const currentTime = Date.now();
    //     if (currentTime - this.lastColorChangeTime > this.colorChangeInterval) {
    //       this.lastColorChangeTime = currentTime;
    //       this.colorIndex = (this.colorIndex + 1) % this.cycleColors.length;
    //       const hexColor = `#${this.cycleColors[this.colorIndex].map(c => c.toString(16).padStart(2, '0')).join('')}`;
    //       console.log(chalk.bgHex(hexColor).white(` Color cycled to index ${this.colorIndex}, color ${hexColor} `));
    //       if (!this.debug) {
    //           this.strip.setColourRGB(this.cycleColors[this.colorIndex][0], this.cycleColors[this.colorIndex][1], this.cycleColors[this.colorIndex][2]);
    //       }
    //   }
    // }
    colorCycleMode() {
      this.colorSet = false;
      const currentTime = Date.now();
      if (currentTime - this.lastColorChangeTime > this.colorChangeInterval) {
          this.lastColorChangeTime = currentTime;
          const startColor = this.cycleColors[this.colorIndex];
          this.colorIndex = (this.colorIndex + 1) % this.cycleColors.length;
          const endColor = this.cycleColors[this.colorIndex];
          const duration = 5000; // Duration of color transition in milliseconds

          const transitionInterval = 50; // Interval for color transition in milliseconds
          const steps = duration / transitionInterval;
          let step = 0;

          const transition = setInterval(() => {
              step++;
              const ratio = step / steps;
              const transitionColor = startColor.map((channel, index) => {
                  const delta = endColor[index] - channel;
                  return Math.round(channel + delta * ratio);
              });
              if (this.mode === this.MODES.COLOR_CYCLE) {
                const hexColor = `#${transitionColor.map(c => c.toString(16).padStart(2, '0')).join('')}`;
                console.log(chalk.bgHex(hexColor).white(` Color transitioned to index ${this.colorIndex}, color ${hexColor} `));
                if (!this.debug) {
                    this.strip.setColourRGB(transitionColor[0], transitionColor[1], transitionColor[2]);
                }
              }

              if (step >= steps) {
                  clearInterval(transition);
              }
          }, transitionInterval);
      }
  }

    breathingMode() {
      this.colorSet = false;
      const currentTime = Date.now();
      const timeIntoCycle = (currentTime % this.breathCycleDuration) / this.breathCycleDuration;
      const brightness = (Math.sin(timeIntoCycle * 2 * Math.PI) + 1) / 2;
      const adjustedColor = this.cycleColors[0].map(c => Math.round(c * brightness));
      const hexColor = `#${adjustedColor.map(c => c.toString(16).padStart(2, '0')).join('')}`;
      console.log(chalk.bgHex(hexColor).white(` Breathing mode: brightness ${brightness.toFixed(2)}, color ${hexColor} `));
      if (!this.debug) {
          this.strip.setColourRGB(adjustedColor[0], adjustedColor[1], adjustedColor[2]);
      }
    }

    run() {
        this.interval = setInterval(() => {
            // console.log(`Current mode: ${this.mode === this.MODES.STATIC_COLOR ? 'STATIC_COLOR' : this.mode === this.MODES.COLOR_CYCLE ? 'COLOR_CYCLE' : 'BREATHING'}.`);
            switch (this.mode) {
                case this.MODES.STATIC_COLOR:
                    this.staticColorMode();
                    break;
                case this.MODES.COLOR_CYCLE:
                    this.colorCycleMode();
                    break;
                case this.MODES.BREATHING:
                    this.breathingMode();
                    break;
            }
        }, 100);
    }

    cleanup() {
        console.log('Cleaning up resources...');
        if (!this.debug) {
            try {
                clearInterval(this.interval);
                this.strip.setColourOff();
                this.strip.cleanup();
            } catch (error) {
                console.error('Error clearing interval or cleaning up LEDStrip.');
            }
        }
    }
}

module.exports = LEDController;
